cowprotocol/services
# Hot reload log filter

# Description
To reduce infra cost we try to be mindful about which logs we enable. But sometimes logs that are not useful most of the time could help debug an issue (checking what exact requests / responses a particular solver sees comes to mind).
So far we always had to restart the whole backend stack only to adjust a tiny log filter. This can be quite problematic on prod because every restart is a tiny outage due to the native price cache needing to be populated again.

# Changes
Initially I thought about exposing an API endpoint that allows for this but that would be a change that has to happen in every single process and some binaries currently don't even expose any endpoints.
The second idea was to implement a signal handler that when triggered reloads the contents of a file. This works and was pretty straight forward but be a bit awkward in practice because you first have to create the file and then send the signal (e.g. with `htop`).

Finally I ended up implementing with a UNIX socket where anyone with access to the cluster can connect to with `nc`. That not only is more convenient to apply the change but we also get immediate feedback instead of having to monitor logs in kibana for your change. Also even if you filter out all logs completely you would still get responses in the socket.

## How to test
Run any binary that uses the `observe` crate (which I think are all of them) and then connect to the socket via `nc -U /tmp/log_filter_override_<program_name>_<pid>.sock` and write your preferred filter into it.

If you are done with your changes you can send `reset` any time (even after dropping the connection) to revert back to the log filters we started the program with.

The very verbose socket name was only used to support the use case where you run multiple identical processes (e.g. driver) locally on your machine. On our cluster there should always be exactly a single process running per pod so we could have gone with a simple name like `/tmp/log_filter_override.sock` there.

<img width="990" alt="Screenshot 2024-06-28 at 19 08 49" src="https://github.com/cowprotocol/services/assets/19190235/64324d1d-b63b-41ad-b117-96729bf5e59b">
